### DISCUSSION TITLE
AssertionError: vivado_hls not found in PATH

### DISCUSSION THREAD

**Discussion by MayMyat-THU** [Xilinx/finn#518] [Q&A] [answered]
Hi, I'm trying to run hardware generation step in "cnv_end2end_example" from bnn-pynq of finn repo and I am having this error;

Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/opt/conda/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/workspace/finn/src/finn/transformation/fpgadataflow/hlssynth_ip.py", line 69, in applyNodeLocal
    inst.ipgen_singlenode_code()
  File "/workspace/finn/src/finn/custom_op/fpgadataflow/hlscustomop.py", line 323, in ipgen_singlenode_code
    builder.build(code_gen_dir)
  File "/workspace/finn-base/src/finn/util/hls.py", line 62, in build
    assert which(self.backend) is not None, "%s not found in PATH" % self.backend
AssertionError: vivado_hls not found in PATH

I have already assigned FINN_XILINX_PATH and FINN_XILINX_VERSION according to the finn.readthedocs.io
I'm on Ubuntu 20.04 and my Vivado is 2021.2. 

I assume that I need to assign Vivado installation path but how do I do it? I'm new to Finn and docker so if there's anyone who knows how to do it, please tell me. Thanks in advanced.

**Comment by shashwat1198** [comment]
Can you share the FINN_XILINX_PATH that you have currently exported? 
It would be the directory where your Vivado is installed.

In addition to these two I also export the settings64.sh file of my vivado installation like below:
{export VIVADO_PATH={Vivado Installation directory}/Vivado/2019.1/settings64.sh}

You can also look at the system requirements here : https://finn.readthedocs.io/en/latest/getting_started.html#system-requirements
Because I don't know if the 2021.1 version of vivado is supported yet.

**Reply by MayMyat-THU** [reply]
Sorry for the delay. I checked the link you've mentioned and I doubted that I don't fit into their system requirements. So, I restarted from the beginning and now I'm currently on Ubuntu 18.04.3 and Vivado 2020.1 (as described in https://finn.readthedocs.io/en/latest/getting_started.html#system-requirements )
Unfortunately I am still having the same error.

I didn't change any variable in the script "run-docker.sh". Instead, I set up the path in ~/etc/profile as 

export FINN_XILINX_PATH=/tools/Xilinx
export FINN_XILINX_VERSION= 2020.1
export VIVADO_PATH=/tools/Xilinx/Vivado/2020.1/settings64.sh

Is it the correct way to set Vivado path? Or should I assign FINN_XILINX_PATH and FINN_XILINX_VERSION inside the "run-docker.sh" script itself? Thanks it advanced.

**Reply by caoimhelynch** [reply]
Hi @MayMyat-THU did you ever get a solution for this? I am currently running into the same issue

**Reply by YangNuoCheng** [reply]
If we set **FINN_XILINX_PATH=(/tools/Xilinx) ** and **FINN_XILINX_VERSION(2020.1) ** in **run-docker.sh**, the file can get the right PATH but I faced more questions.(Cant mount /tools/Xilinx read only file system)

**Comment by bestamigunay** [comment]
You can try `feature/vitis_hls_2021.2` branch. And you must be sure about that you also installed Vitis HLS with Vivado.

**Reply by MayMyat-THU** [reply]
I tried it but it didn't work. Seems like FINN is only compatible with Vivado 2020.1 and below.

**Reply by MehdiRh17** [reply]
I have the same problem with 2019.1

**Reply by troibe** [reply]
I tried the vitis_hls branch for 2020.2
While I don't run into the "vivado_hls not found in PATH" error now I run into `ERROR: [BD 5-390] IP definition not found for VLNV: xilinx.com:hls:Thresholding_Batch_0:1.0`.
@bestamigunay did you run into this too?
If so did you find a way to fix it? I guess it fails at creating a Wrapper for the pynq board.

**Reply by troibe** [reply]
Ok so apparently that was caused by the Y22k bug. After applying this [patch](https://support.xilinx.com/s/article/76960?language=en_US ) I could make some more progress.
Now I'm running into the following issue:
`ERROR: [BD 5-106] Arguments to the connect_bd_intf_net command cannot be empty.`

**Comment by nhma20** [comment, marked as answer]
Have you sourced the right Vivado version? E.g. "source /tools/Xilinx/Vivado/2020.1/settings64.sh"
My VIVADO_PATH looks different to yours: "export VIVADO_PATH=/tools/Xilinx/Vivado/2020.2"

For newer versions of Xilinx tools, I use Vivado 2020.2 which works with the "feature/vitis_hls" branch of FINN as far as I can tell.

**Comment by bansalrajesh** [comment]
Has anyone found the solution of this problem. 
if correct FINN_XILINX_PATH is set ...it has to be one level before version Finally I  saw in docker  Found Vitis at
Found Vitis_HLS at ...but the PATH settings are not updated ...i don't know if that's issue ...

**Comment by CNStanLee** [comment]
I met this again today. Here is the note.
This is probably caused by the path setting for a normal user, but Docker is running as root.
So, you can do this:
1. Create a file named "start_finn.sh" in the same path as run-docker.sh, and input the code below into it:
export FINN_XILINX_PATH=/tools/Xilinx
export FINN_XILINX_VERSION=2022.2
source /tools/Xilinx/Vivado/2022.2/settings64.sh
export VIVADO_PATH=/tools/Xilinx/Vivado/2022.2
export JUPYTER_PORT=8886
export NVIDIA_VISIBLE_DEVICES=0
./run-docker.sh notebook
2. Modify the path to yours
3. Run a terminal in FINN project path, then input "sudo su"
4. Run "source start_finn.sh"
